packages/mobile/src/components/core/PullToRefresh.tsx

```diff
@@ -0,0 +1,256 @@
+  const [isDebouncing, setIsDebouncing] = useState(false)
+
+  const timeoutRef = useRef<NodeJS.Timeout>()
+  const handleRefresh = useCallback(() => {
+    onRefresh?.()
+    setIsDebouncing(true)
+    timeoutRef.current = setTimeout(() => {
+      setIsDebouncing(false)
+    }, DEBOUNCE_TIME_MS)
+  }, [onRefresh])
+
+  const scrollTo = useCallback(
+    (y: number) => {
+      if (scrollResponder && 'scrollTo' in scrollResponder) {
+        scrollResponder?.scrollTo({ y, animated: true })
+      }
+      if (scrollResponder && 'scrollToOffset' in scrollResponder) {
+        scrollResponder?.scrollToOffset({ offset: y, animated: true })
+      }
+    },
+    [scrollResponder]
+  )
+
+  useEffect(() => {
+    if (!isRefreshing && !isDebouncing) {
+      scrollTo(0)
+    }
+  }, [isRefreshing, isDebouncing, scrollTo])
+
+  const onScrollBeginDrag = useCallback(() => {
+    setIsMomentumScroll(false)
+  }, [setIsMomentumScroll])
+
+  const onScrollEndDrag = useCallback(() => {
+    setIsMomentumScroll(true)
+    if (isRefreshing) {
+      scrollTo(-50)
+    }
+  }, [setIsMomentumScroll, scrollTo, isRefreshing])
+
+  const onScroll = attachToScroll(scrollAnim)
+
+  return {
+    isRefreshing: isRefreshing || isDebouncing,
+    isRefreshDisabled: isMomentumScroll,
+    handleRefresh,
+    scrollAnim,
+    onScroll,
+    onScrollBeginDrag,
+    onScrollEndDrag
+  }
+}
+
+type PullToRefreshProps = {
+  isRefreshing?: boolean
+  onRefresh?: () => void
+  scrollAnim?: Animated.Value
+  isRefreshDisabled?: boolean
+  topOffset?: number
+  color?: string
+}
+
+/**
+ * Custom pull to refresh to be used with React Native ScrollViews
+ * (FlatList, SectionList, etc.).
+ *
+ * Usage:
+ * ```
+ * const scrollAnim = useRef(new Animated.Value(0)).current
+ *
+ * const MyList = () => {
+ *  return (
+ *    <View>
+ *      <PullToRefresh
+ *        isRefreshing={}
+ *        onRefresh={}
+ *      />
+ *      <ScrollView
+ *        onScroll={attachToScroll(scrollAnim)}
+ *        {...props}
+ *      />
+ *  )
+ * }
+ * ```
+ *
+ * Two usage modes of PullToRefresh are provided.
+ * If the PullToRefresh is intended to appear in the overscroll
+ * space above the view, see `useOverscrollHandlers`
+ * Otherwise, the component may suffice.
+ */
+export const PullToRefresh = ({
+  isRefreshing,
+  onRefresh,
+  isRefreshDisabled,
+  scrollAnim,
+  topOffset = 0,
+  color
+}: PullToRefreshProps) => {
+  const styles = useStyles(topOffset)()
+  const { neutralLight4 } = useThemeColors()
+
+  const [didHitTop, setDidHitTop] = useState(false)
+  const hitTop = useRef(false)
+
+  const [shouldShowSpinner, setShouldShowSpinner] = useState(false)
+  const animationRef = useRef<LottieView | null>()
+
+  const icon = shouldShowSpinner ? IconRefreshSpin : IconRefreshPull
+  const colorizedIcon = useMemo(
+    () =>
+      colorize(icon, {
+        // arrow Outlines 4.Group 1.Stroke 1
+        'assets.0.layers.0.shapes.0.it.1.c.k': color || neutralLight4,
```

Comment:
This is dope but the specificity here is giving me anxiety

Reply:
The good news is you don't have to write this by hand!!!
https://colorize-react-native-lottie.netlify.app/